x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/date/DateDiff.java

```diff
@@ -213,33 +220,106 @@ Expression endTimestamp() {
+    static int processNanosMillis(BytesRef unit, long startTimestamp, long endTimestamp) throws IllegalArgumentException {
+        return processNanosMillis(Part.resolve(unit.utf8ToString()), startTimestamp, endTimestamp);
+    }
+
+    @Evaluator(extraName = "ConstantMillisNanos", warnExceptions = { IllegalArgumentException.class, InvalidArgumentException.class })
+    static int processMillisNanos(@Fixed Part datePartFieldUnit, long startTimestamp, long endTimestamp) throws IllegalArgumentException {
+        ZonedDateTime zdtStart = ZonedDateTime.ofInstant(Instant.ofEpochMilli(startTimestamp), UTC);
+        ZonedDateTime zdtEnd = ZonedDateTime.ofInstant(DateUtils.toInstant(endTimestamp), UTC);
+        return datePartFieldUnit.diff(zdtStart, zdtEnd);
+    }
+
+    @Evaluator(extraName = "MillisNanos", warnExceptions = { IllegalArgumentException.class, InvalidArgumentException.class })
+    static int processMillisNanos(BytesRef unit, long startTimestamp, long endTimestamp) throws IllegalArgumentException {
+        return processMillisNanos(Part.resolve(unit.utf8ToString()), startTimestamp, endTimestamp);
+    }
+
+    @FunctionalInterface
+    public interface DateDiffFactory {
+        ExpressionEvaluator.Factory build(
+            Source source,
+            ExpressionEvaluator.Factory unitsEvaluator,
+            ExpressionEvaluator.Factory startTimestampEvaluator,
+            ExpressionEvaluator.Factory endTimestampEvaluator
+        );
+    }
+
+    @FunctionalInterface
+    public interface DateDiffConstantFactory {
+        ExpressionEvaluator.Factory build(
+            Source source,
+            Part unitsEvaluator,
+            ExpressionEvaluator.Factory startTimestampEvaluator,
+            ExpressionEvaluator.Factory endTimestampEvaluator
+        );
     }
 
     @Override
     public ExpressionEvaluator.Factory toEvaluator(ToEvaluator toEvaluator) {
+        if (startTimestamp.dataType() == DATETIME && endTimestamp.dataType() == DATETIME) {
+            return toEvaluator(toEvaluator, DateDiffConstantMillisEvaluator.Factory::new, DateDiffMillisEvaluator.Factory::new);
+        } else if (startTimestamp.dataType() == DATE_NANOS && endTimestamp.dataType() == DATE_NANOS) {
+            return toEvaluator(toEvaluator, DateDiffConstantNanosEvaluator.Factory::new, DateDiffNanosEvaluator.Factory::new);
+        } else if (startTimestamp.dataType() == DATE_NANOS && endTimestamp.dataType() == DATETIME) {
+            return toEvaluator(toEvaluator, DateDiffConstantNanosMillisEvaluator.Factory::new, DateDiffNanosMillisEvaluator.Factory::new);
+        } else if (startTimestamp.dataType() == DATETIME && endTimestamp.dataType() == DATE_NANOS) {
+            return toEvaluator(toEvaluator, DateDiffConstantMillisNanosEvaluator.Factory::new, DateDiffMillisNanosEvaluator.Factory::new);
+        }
+        throw new UnsupportedOperationException("How'd we get here?");
```

Comment:
This is technically impossible as long as the resolveType and the rules/ESQL is right, but I would show the `startTimestamp` and `endTimestamp` types in the error, just in case